src/Solcore/Frontend/TypeInference/SccAnalysis.hs

```diff
@@ -38,28 +38,44 @@ sccAnalysis' :: CompUnit Name -> SCC (CompUnit Name)
 sccAnalysis' (CompUnit imps ds)
   = do
       cs' <- mapM sccContract cs
-      CompUnit imps <$> analysis (cs' ++ ds')
+      CompUnit imps <$> analysis group (cs' ++ ds')
     where
       isContract (TContr _) = True
       isContract _ = False
 
+      isFunDef (TFunDef _) = True
+      isFunDef _ = False
+
+      group xs =
+        if (all isFunDef xs && length xs > 1)
+        then singleton (TMutualDef xs)
+        else xs
+
       (cs, ds') = partition isContract ds
 
 -- sort inner contract definitions
 
 sccContract :: TopDecl Name -> SCC (TopDecl Name)
 sccContract (TContr (Contract n vs ds))
-  = (TContr . Contract n vs) <$> analysis ds
+  = (TContr . Contract n vs) <$> analysis group ds
+    where
+      group xs =
+        if (all isFunDef xs && length xs > 1)
+        then singleton (CMutualDecl xs)
+        else xs
+
+      isFunDef (CFunDecl _) = True
+      isFunDef _ = False
 sccContract d = pure d
 
-analysis :: (Ord a, Names a, Decl a) => [a] -> SCC [a]
-analysis ds
+analysis :: (Ord a, Names a, Decl a, Show a) => ([a] -> [a]) -> [a] -> SCC [a]
```

Comment:
we couldn't find a clean way to make this generic, so ended up implementing `group` twice with concrete types and passing it in. sugggestions welcome.